File: scripts/tasks/apex_put_files.mjs
/**
 * Upload files to Apex (Pterodactyl) when Python is unavailable.
 * Secrets: repo-root .apex_deploy.json (never commit).
 *
 *   node scripts/tasks/apex_put_files.mjs --put local:remote ... [--delete remote] [--restart]
 */
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "../..");
const secrets = JSON.parse(fs.readFileSync(path.join(ROOT, ".apex_deploy.json"), "utf8"));
const panel = String(secrets.apex_panel || "https://panel.apexnodes.xyz").replace(/\/$/, "");
const sid = secrets.apex_server_id;
const key = secrets.apex_api_key;
if (!sid || !key) {
  console.error("missing apex_server_id / apex_api_key");
  process.exit(1);
}

const headers = {
  Authorization: `Bearer ${key}`,
  Accept: "Application/vnd.pterodactyl.v1+json",
};

function parseArgs(argv) {
  const puts = [];
  const deletes = [];
  let restart = false;
  for (let i = 0; i < argv.length; i++) {
    const a = argv[i];
    if (a === "--restart") restart = true;
    else if (a === "--put") {
      const spec = argv[++i];
      const idx = spec.indexOf(":");
      if (idx < 1) throw new Error(`bad --put ${spec}`);
      puts.push({ local: spec.slice(0, idx), remote: spec.slice(idx + 1) });
    } else if (a === "--delete") deletes.push(argv[++i]);
    else throw new Error(`unknown arg ${a}`);
  }
  return { puts, deletes, restart };
}

async function api(method, p, body, extra = {}) {
  const url = `${panel}${p}`;
  const init = { method, headers: { ...headers, ...extra.headers } };
  if (body !== undefined) init.body = body;
  const res = await fetch(url, init);
  const text = await res.text();
  if (!res.ok) throw new Error(`${method} ${p} HTTP ${res.status}: ${text.slice(0, 400)}`);
  return text ? JSON.parse(text) : {};
}

async function writeFile(remote, localAbs) {
  const posix = "/" + remote.replace(/\\/g, "/").replace(/^\/+/, "");
  const slash = posix.lastIndexOf("/");
  const dir = slash <= 0 ? "/" : posix.slice(0, slash);
  const name = posix.slice(slash + 1);
  const buf = fs.readFileSync(localAbs);
  // Panel /files/write is nginx-capped (~1MB). Jars go through Wings signed upload.
  if (buf.length < 512 * 1024) {
    const q = `file=${encodeURIComponent(posix)}`;
    await api("POST", `/api/client/servers/${sid}/files/write?${q}`, buf, {
      headers: { "Content-Type": "application/octet-stream" },
    });
    console.log(`put ${posix} (${buf.length} bytes)`);
    return;
  }
  const signed = await api(
    "GET",
    `/api/client/servers/${sid}/files/upload?directory=${encodeURIComponent(dir)}`,
  );
  const rawUrl = signed?.attributes?.url;
  if (!rawUrl) throw new Error("no signed upload url");
  const u = new URL(rawUrl);
  u.searchParams.set("directory", dir);
  const fd = new FormData();
  fd.append("files", new Blob([buf]), name);
  const res = await fetch(u.toString(), { method: "POST", body: fd });
  const text = await res.text();
  if (!res.ok) throw new Error(`upload ${posix} HTTP ${res.status}: ${text.slice(0, 400)}`);
  console.log(`put ${posix} (${buf.length} bytes)`);
}

async function deleteFiles(files) {
  if (!files.length) return;
  const names = files.map((f) => f.replace(/\\/g, "/").replace(/^\/+/, ""));
  await api(
    "POST",
    `/api/client/servers/${sid}/files/delete`,
    JSON.stringify({ root: "/", files: names }),
    { headers: { "Content-Type": "application/json" } },
  );
  for (const n of names) console.log(`deleted /${n}`);
}

async function command(cmd) {
  try {
    await api(
      "POST",
      `/api/client/servers/${sid}/command`,
      JSON.stringify({ command: cmd }),
      { headers: { "Content-Type": "application/json" } },
    );
    console.log(`console: ${cmd}`);
  } catch (e) {
    console.warn(`console skip: ${e.message}`);
  }
}

async function power(signal) {
  await api(
    "POST",
    `/api/client/servers/${sid}/power`,
    JSON.stringify({ signal }),
    { headers: { "Content-Type": "application/json" } },
  );
  console.log(`power ${signal}`);
}

async function waitRunning(timeoutMs = 180000) {
  const deadline = Date.now() + timeoutMs;
  let seenDown = false;
  while (Date.now() < deadline) {
    const data = await api("GET", `/api/client/servers/${sid}/resources`);
    const state = data?.attributes?.current_state || "unknown";
    console.log(`  panel ${state}`);
    if (state === "starting" || state === "offline" || state === "stopping") seenDown = true;
    if (seenDown && state === "running") return;
    await new Promise((r) => setTimeout(r, 5000));
  }
  console.warn("WARN not confirmed running");
}

const { puts, deletes, restart } = parseArgs(process.argv.slice(2));
for (const { local, remote } of puts) {
  const abs = path.isAbsolute(local) ? local : path.join(ROOT, local);
  if (!fs.existsSync(abs)) throw new Error(`missing ${abs}`);
  await writeFile(remote, abs);
}
await deleteFiles(deletes);
if (restart) {
  await command("save-all");
  await new Promise((r) => setTimeout(r, 3000));
  await power("restart");
  await waitRunning();
}

File: scripts/tasks/apex_put_files.py
#!/usr/bin/env python3
"""Upload/delete files on the Apex panel via the Pterodactyl files API.

Fallback for when SFTP auth/connection is flaky.

Usage:
  python scripts/tasks/apex_put_files.py upload mods server/mods/aqualumen-forge-1.20.1-0.3.55-alpha.jar
  python scripts/tasks/apex_put_files.py delete mods aqualumen-forge-1.20.1-0.3.53-alpha.jar
"""
from __future__ import annotations

import json
import os
import sys
import urllib.error
import urllib.request
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
SECRETS = ROOT / ".apex_deploy.json"

PANEL = os.environ.get("AQUATECH_APEX_PANEL", "https://panel.apexnodes.xyz").rstrip("/")
SERVER_ID = os.environ.get("AQUATECH_APEX_SERVER_ID", "")
API_KEY = os.environ.get("AQUATECH_APEX_API_KEY", "")


def load_secrets() -> None:
    global PANEL, SERVER_ID, API_KEY
    if SECRETS.is_file():
        data = json.loads(SECRETS.read_text(encoding="utf-8"))
        PANEL = str(data.get("apex_panel") or PANEL).rstrip("/")
        SERVER_ID = str(data.get("apex_server_id") or SERVER_ID)
        API_KEY = str(data.get("apex_api_key") or API_KEY)
    if not SERVER_ID or not API_KEY:
        sys.exit("missing apex_server_id / apex_api_key (env or .apex_deploy.json)")


def api(method: str, path: str, data: bytes | None = None, content_type: str | None = None) -> tuple[int, bytes]:
    headers = {
        "Authorization": f"Bearer {API_KEY}",
        "Accept": "application/json",
        "User-Agent": "AquaTechPanelUploader",
    }
    if content_type:
        headers["Content-Type"] = content_type
    req = urllib.request.Request(f"{PANEL}{path}", data=data, headers=headers, method=method)
    try:
        with urllib.request.urlopen(req, timeout=600) as resp:
            return resp.status, resp.read()
    except urllib.error.HTTPError as e:
        body = e.read()
        print(f"HTTP {e.code} {path}: {body[:400].decode('utf-8', 'replace')}", file=sys.stderr)
        return e.code, body


def upload(remote_dir: str, local_path: Path) -> int:
    directory = "/" + remote_dir.strip("/")
    target = local_path.name
    boundary = "----AquaTechPanelUploadBoundary7f3c9a"
    body = b"".join([
        f"--{boundary}\r\n".encode(),
        f'Content-Disposition: form-data; name="file"; filename="{target}"\r\n'.encode(),
        b"Content-Type: application/octet-stream\r\n\r\n",
        local_path.read_bytes(),
        f"\r\n--{boundary}--\r\n".encode(),
    ])
    status, resp = api("POST", f"/api/client/servers/{SERVER_ID}/files/write?directory={directory}",
                      data=body, content_type=f"multipart/form-data; boundary={boundary}")
    if 200 <= status < 300:
        print(f"OK uploaded {target} -> {directory} ({local_path.stat().st_size} bytes)")
        return 0
    if resp:
        print(resp[:200].decode("utf-8", "replace"), file=sys.stderr)
    return 1


def delete(remote_dir: str, names: list[str]) -> int:
    root = "/" + remote_dir.strip("/")
    payload = json.dumps({"root": root, "files": names}).encode()
    status, _ = api("POST", f"/api/client/servers/{SERVER_ID}/files/delete",
                    data=payload, content_type="application/json")
    if 200 <= status < 300:
        print(f"OK deleted from {root}: {', '.join(names)}")
        return 0
    return 1


def pull(remote_dir: str, url: str) -> int:
    """Ask the panel to download a file by URL into remote_dir (bypasses body limits)."""
    directory = "/" + remote_dir.strip("/")
    payload = json.dumps({"url": url, "directory": directory}).encode()
    status, _ = api("POST", f"/api/client/servers/{SERVER_ID}/files/pull",
                    data=payload, content_type="application/json")
    if 200 <= status < 300:
        print(f"OK panel pulls {url} -> {directory}")
        return 0
    return 1


def main() -> int:
    load_secrets()
    if len(sys.argv) < 4:
        print(__doc__)
        return 2
    action, remote_dir = sys.argv[1], sys.argv[2]
    if action == "upload":
        rc = 0
        for arg in sys.argv[3:]:
            p = Path(arg)
            if not p.is_file():
                print("missing local file:", p, file=sys.stderr)
                rc = 1
                continue
            rc |= upload(remote_dir, p)
        return rc
    if action == "delete":
        return delete(remote_dir, sys.argv[3:])
    if action == "pull":
        rc = 0
        for url in sys.argv[3:]:
            rc |= pull(remote_dir, url)
        return rc
    print("unknown action:", action)
    return 2


if __name__ == "__main__":
    raise SystemExit(main())
